src/ziggurat/messaging/consumer.clj

```diff
@@ -8,25 +8,43 @@
             [taoensso.nippy :as nippy]
             [ziggurat.config :refer [get-in-config]]
             [ziggurat.messaging.connection :refer [connection]]
+            [ziggurat.middleware.default :as mw]
             [ziggurat.sentry :refer [sentry-reporter]]
             [ziggurat.messaging.util :refer :all]
             [ziggurat.metrics :as metrics]
-            [ziggurat.util.error :refer [report-error]]))
+            [ziggurat.util.error :refer [report-error]])
+  (:import (com.ziggurat.proto MessagePayloadProto$MessagePayload)
+           (com.google.protobuf ByteString)))
+
+(defn- nippy-deserialize
+  [frozen]
+  (try
+    (let [thaw (nippy/thaw frozen)]
+      (into {} thaw))                                       ;Converting defrecord into map, since proto does not understand defrecord
```

Comment:
Why are we creating a map. Since the code is freezing a map, when thawed, it would always be a map

I saw the comment but nippy messages are never meant to go to proto

Am I missing something here?

Reply:
Scenario:
Version < 4.2.0 sends the message to rabbitmq for retry thereby sending in a nippy serialised message payload.
A deployment happens.
Version 4.2.0 picks up the message, deserializes the message to message payload and probably needs to serialize it again (to be sent to delay queue or deadset), but this time it happens via proto which would fail, causing an exception and the message is lost.

Converting it to a map solves this case.

Reply:
hmm. okay. 

I initially thought what could be the problem with a defrecord map but I read that defrecord generates a class object. This causes proto serialization to fail.